modules/systemd-journal/journal-reader.c

```diff
@@ -193,18 +174,50 @@ __handle_data(gchar *key, gchar *value, gpointer user_data)
     {
       msg->pri = (msg->pri & ~7) | atoi(value);
     }
+}
+
+static void
+__fill_message(JournalReaderOptions *options, LogMessage *msg, gchar *key, gchar *value, gssize value_len)
+{
+  if (!options->prefix)
+    {
+      log_msg_set_value_by_name(msg, key, value, value_len);
+    }
   else
     {
-      if (!options->prefix)
-        {
-          log_msg_set_value_by_name(msg, key, value, value_len);
-        }
-      else
-        {
-          gchar *prefixed_key = g_strdup_printf("%s%s", options->prefix, key);
-          log_msg_set_value_by_name(msg, prefixed_key, value, value_len);
-          g_free(prefixed_key);
-        }
+      gchar *prefixed_key = g_strdup_printf("%s%s", options->prefix, key);
+      log_msg_set_value_by_name(msg, prefixed_key, value, value_len);
+      g_free(prefixed_key);
+    }
+}
+
+static void
+__handle_data(gchar *key, gchar *value, gpointer user_data)
+{
+  gpointer *args = user_data;
+
+  LogMessage *msg = args[0];
+  JournalReaderOptions *options = args[1];
+  gssize value_len = MIN(strlen(value), options->max_field_size);
+
+  __map_key_value_pairs_to_syslog_macros(msg, key, value, value_len);
+  __fill_message(options, msg, key, value, value_len);
+}
+
+static void
+__set_program(LogMessage *msg)
+{
+  gssize value_length = 0;
+  const gchar *value = log_msg_get_value_by_name(msg, "SYSLOG_IDENTIFIER", &value_length);
+
```

Comment:
prefix is added by _fill_message().

I think the code is cleaner without the nvhandle-s. I would optimize it further when somebody reports back that it is much slower.